### PR TITLE
UCT/IB/MLX5: DV UAR alloc type NC support

### PR DESCRIPTION
**Pull request by Artemy-Mellanox**

### CODE REVIEW

File: src/uct/ib/mlx5/ib_mlx5.c [hunk 1]
```diff
@@ -405,8 +405,16 @@ ucs_status_t uct_ib_mlx5_devx_uar_init(uct_ib_mlx5_devx_uar_t *uar,
...
+    unsigned flags      = UCT_IB_MLX5_UAR_ALLOC_TYPE_WC;
+
+retry:
+    uar->uar            = mlx5dv_devx_alloc_uar(md->super.dev.ibv_context,
+                                                flags);
     if (uar->uar == NULL) {
+        if (flags != UCT_IB_MLX5_UAR_ALLOC_TYPE_NC) {
+            flags = UCT_IB_MLX5_UAR_ALLOC_TYPE_NC;
+            goto retry;
+        }
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.c:417]: IMO, calling mlx5dv_devx_alloc_uar twice(instead of goto/retry) would be more readable

Comment (Artemy-Mellanox) [reply src/uct/ib/mlx5/ib_mlx5.c:417]: don't we want to cache the failure?

File: src/uct/ib/mlx5/ib_mlx5.c [hunk 2]
```diff
@@ -405,10 +405,20 @@ ucs_status_t uct_ib_mlx5_devx_uar_init(uct_ib_mlx5_devx_uar_t *uar,
                                        uct_ib_mlx5_mmio_mode_t mmio_mode)
 {
 #if HAVE_DEVX
-    uar->uar            = mlx5dv_devx_alloc_uar(md->super.dev.ibv_context, 0);
+    uar->uar            = mlx5dv_devx_alloc_uar(md->super.dev.ibv_context,
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.c:408]: pls remove indentation where not needed

File: src/uct/ib/mlx5/ib_mlx5.c [hunk 3]
```diff
@@ -405,10 +405,20 @@ ucs_status_t uct_ib_mlx5_devx_uar_init(uct_ib_mlx5_devx_uar_t *uar,
...
+                                                UCT_IB_MLX5_UAR_ALLOC_TYPE_WC);
     if (uar->uar == NULL) {
-        ucs_error("mlx5dv_devx_alloc_uar() failed: %m");
-        return UCS_ERR_NO_MEMORY;
+        if (UCT_IB_MLX5_UAR_ALLOC_TYPE_WC == UCT_IB_MLX5_UAR_ALLOC_TYPE_NC) {
+            ucs_error("mlx5dv_devx_alloc_uar(WC) failed: %m");
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.c:412]: 1. print a debug message if WC failed and we're doing fallback
2. print actual flag value, ucs_error("mlx5dv_devx_alloc_uar(WC, flags=0x%x) failed: %m", UCT_IB_MLX5_UAR_ALLOC_TYPE_WC);

File: src/uct/ib/mlx5/ib_mlx5.c [hunk 4]
```diff
@@ -405,10 +405,20 @@ ucs_status_t uct_ib_mlx5_devx_uar_init(uct_ib_mlx5_devx_uar_t *uar,
...
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        uar->uar        = mlx5dv_devx_alloc_uar(md->super.dev.ibv_context,
+                                                UCT_IB_MLX5_UAR_ALLOC_TYPE_NC);
+        if (uar->uar == NULL) {
+            ucs_error("mlx5dv_devx_alloc_uar(NC) failed: %m");
+            return UCS_ERR_NO_MEMORY;
+        }
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.c:421]: IMO, `#define UCT_IB_MLX5_UAR_ALLOC_TYPE_NC   0` is not correct (but only `#define UCT_IB_MLX5_UAR_ALLOC_TYPE_WC   0` is correct)
maybe add `#ifdef HAVE_DECL_MLX5DV_UAR_ALLOC_TYPE_NC` here?
and check for HAVE_DECL_MLX5DV_UAR_ALLOC_TYPE_WC separately in configure?

File: src/uct/ib/mlx5/ib_mlx5.c [hunk 5]
```diff
@@ -405,20 +405,23 @@ ucs_status_t uct_ib_mlx5_devx_uar_init(uct_ib_mlx5_devx_uar_t *uar,
...
     if (uar->uar == NULL) {
-        if (UCT_IB_MLX5_UAR_ALLOC_TYPE_WC == UCT_IB_MLX5_UAR_ALLOC_TYPE_NC) {
-            ucs_error("mlx5dv_devx_alloc_uar(WC) failed: %m");
-            return UCS_ERR_NO_MEMORY;
-        }
-
-        uar->uar        = mlx5dv_devx_alloc_uar(md->super.dev.ibv_context,
-                                                UCT_IB_MLX5_UAR_ALLOC_TYPE_NC);
+#if HAVE_DECL_MLX5DV_UAR_ALLOC_TYPE_NC
+        ucs_debug("UAR alloc WC failed. Fallback to NC\n");
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.c:412]: ucs_debug("mlx5dv_devx_alloc_uar(device=%s, flags=0x%x) failed: %m, fallback to NC\n");

File: src/uct/ib/mlx5/ib_mlx5.c [hunk 6]
```diff
@@ -405,20 +405,23 @@ ucs_status_t uct_ib_mlx5_devx_uar_init(uct_ib_mlx5_devx_uar_t *uar,
...
+        uar->uar = mlx5dv_devx_alloc_uar(md->super.dev.ibv_context,
+                                         UCT_IB_MLX5_UAR_ALLOC_TYPE_NC);
         if (uar->uar == NULL) {
-            ucs_error("mlx5dv_devx_alloc_uar(NC) failed: %m");
+            ucs_error("mlx5dv_devx_alloc_uar(NC, flags=0x%x) failed: %m",
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.c:416]: device=%s

**Comment by yosefe** [comment]
also, revert #6061

File: src/uct/ib/mlx5/ib_mlx5.c [hunk 7]
```diff
@@ -409,16 +409,21 @@ ucs_status_t uct_ib_mlx5_devx_uar_init(uct_ib_mlx5_devx_uar_t *uar,
                                      UCT_IB_MLX5_UAR_ALLOC_TYPE_WC);
     if (uar->uar == NULL) {
 #if HAVE_DECL_MLX5DV_UAR_ALLOC_TYPE_NC
-        ucs_debug("UAR alloc WC failed. Fallback to NC\n");
+        ucs_debug("mlx5dv_devx_alloc_uar(device=%s, flags=0x%x(WC)) "
+                  "failed: %m, fallback to NC",
+                  uct_ib_device_name(&md->super.dev),
+                  UCT_IB_MLX5_UAR_ALLOC_TYPE_WC);
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.c:415]: maybe add function uct_ib_mlx5_devx_alloc_uar(md, flags, err_level=ERROR/DEBUG, char *title="WC"/"NC") which will wrap the debug/error print?

**Review by yosefe — APPROVED** [review]
pls squash

File: src/uct/ib/mlx5/ib_mlx5.c [hunk 8]
```diff
@@ -403,16 +403,22 @@ int uct_ib_mlx5_devx_uar_cmp(uct_ib_mlx5_devx_uar_t *uar,
...
-                "failed: %m%s", uct_ib_device_name(&md->super.dev), flags,
-                title, comment);
+        sprintf(buf, "mlx5dv_devx_alloc_uar(device=%s, flags=0x%x(%s)) "
+                "failed: %m", uct_ib_device_name(&md->super.dev), flags, title);
+        if (fallback == NULL) {
+            ucs_log(log_level, "%s", buf);
+        } else {
+            ucs_log(log_level, "%s, fallback to %s", buf, fallback);
+        }
+
```
Comment (yosefe) [inline src/uct/ib/mlx5/ib_mlx5.c:421]: what was wrong with the previous version?

Comment (Artemy-Mellanox) [reply src/uct/ib/mlx5/ib_mlx5.c:421]: coverity

**Review by yosefe — APPROVED** [review]

**Comment by yosefe** [comment]
@Artemy-Mellanox can u pls squash?

**Pull request merged**